text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+
+The motivation of this proposal is to…
+
+- Reduce the burden and toil for Paketo buildpack maintainer teams
+- Continue publishing dependency updates in a timely and secure manner
+- Decouple installing dependencies from configuring them
+- Separate metadata and the actual dependencies so they can be provided to
+  buildpacks in a number of different and flexible ways
+- Establish a reasonable release schedule for buildpacks that’s based around
+  development, not dependencies and thus enabling buildpacks to support version
+  lines
+- Make it easier to package buildpacks for offline environments.
+
+## Detailed Explanation
+At a high level:
+
+- We will define a metadata format that includes the metadata fields currently
+  present in `buildpack.toml`.
+- We will add a dependency version validation section to `buildpack.toml`
+  metadata, this can be used to state that a buildpack version only supports
+  certain ranges of a given tool, such as Java `11.*` or Node.js `16.*`.
+- Metadata can be provided by anyone. A user can add custom metadata, or source
+  from a third party project. Paketo will provide an official set of metadata
+  against which we will test the Paketo buildpacks. It will be distributed via
+  images in an image registry (Docker Hub).
+- Buildpacks do not care how dependency metadata is distributed, that is a
+  separate concern, instead, they just read metadata from a specified location.
+  It is on the platform to mount the metadata that it fetched at that location.
+- The actual dependencies are accessed via the metadata and that can happen
+  over any protocol (HTTPS/SFTP/FILE) and be distributed in any format
+  (archive/image).
+- Dependency metadata will be removed from `buildpack.toml`
+
+
+## Metadata Format
+The metadata presented to a buildpack will be structured as a directory of
+dependencies.
+
+Like it does now, each dependency will have an id but unlike the present
+situation, the id is namespaced. An id is composed of an organization name and
+a dependency name. It follows the [reverse domain name notation](https://en.wikipedia.org/wiki/Reverse_domain_name_notation) and the
+dependency name is defined as the final item, so `com.example.dep-a` would have
+an organization of `com.example` and a dependency of `dep-a`. It is case
+insensitive, so `dep-a` is no different than `Dep-A`. This is to reduce the
+possibility of [typosquatting](https://en.wikipedia.org/wiki/Typosquatting).
+Allowed characters are [the same as for a valid hostname](https://en.wikipedia.org/wiki/Hostname#Syntax).
+
+The directory structure will contain a folder for each dot-separated segment of
+the dependency’s organization name and in the lowest level directory there will
+be a file named after the dependency name with the extension `toml`, this is
+because the metadata file will be TOML format. For example, with
+`com.example.dep-a`, there would be the following folder structure:
+
+```
+com
+└── example
+    └── dep-a.toml
```

Comment:
```
com
└── example
    └── dep-a
        └── metadata.toml
```

Is there a particular reason to treat namespace and dependency name differently? We could just add one more node to the tree representing the dependency and have a fixed name for the file, e.g. `metadata.toml` or `dependency.toml`.
It doesn't mater much, but would leave room to even extend metadata with additional toml files if we'd ever see a reason to do so.

Alternatively, if we go for a fixed layout with `dep-a.toml` as the leafs, one potential benefit could be to allow the file system structure to be optional. I.e. we could define
```toml
[[com]]
[[com.example]]
[[com.example."dep-a"]]
[[com.example."dep-a".versions]]
name="dep-a"

[[com.example."dep-b"]]
[[com.example."dep-b".versions]]
name="dep-b"
```
to keep things simple in case of a few dependencies and allow the filesystem structure as convenience (nodes in the filesystem leading to TOML table array prefixes) in a parser for dependencies.

Reply:
I like this extensibility. I don't think it costs anything to adopt this over what's proposed in the RFC currently. But I could be missing something.

Reply:
You first proposal is interesting because it would allow us the have N length domain names with very simple backend logic. If you want `com.example.dep-a` or `com.example.group.subgroup.dep-a` you just need to follow the directory path and then grab the metadata file @dmikusa is there any lose in security by doing this?

As for the second proposal I am not sure that I entirely understand the benefit that you are trying to layout. What would be the advantage allowing the filesystem structure to be optional? I feel like it would make things more difficult with you wanted to combine dependency metadata packages together.

Reply:
I think it would make small dependency packages easier to author, think just a small file if it's just a single or two dependencies.

Imagine a user want to consume all dependencies from Paketo, but one of the updates breaks their own code and they need to temporarily substitute metadata (and assets) with an older version.

Reply:
FWIW in the end the dependencies are just a tree with namespace nodes and a named leaf with metadata. 

Merging should be the same, no matter how the tree is represented - even a mixed representation shouldn't make much of a difference.

Reply:
>Is there a particular reason to treat namespace and dependency name differently? We could just add one more node to the tree representing the dependency and have a fixed name for the file, e.g. metadata.toml or dependency.toml.

My thought was that it would keep the directory structure a little flatter, but I don't have a problem with doing it this way. I get what you're saying in terms of flexibility to add more files. I was thinking additional metadata would go in the dependency.toml file, but perhaps there is something else best represented outside of the file. I'm fine with that change, if others agree.

Reply:
>Alternatively, if we go for a fixed layout with dep-a.toml as the leafs, one potential benefit could be to allow the file system structure to be optional.

I can see where you're going, but I'd say in the interest of keeping things MVP let's try things out without this addition and see how it goes first. I suspect there will be tooling to help manage the metadata, so hopefully, it's not too much of a burden. If it turns out to be, we can come back and look at ways to address.